migrate/json_to_terraform.py

```diff
@@ -0,0 +1,340 @@
+import json
+import os
+import sys
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, List, NewType, Set
+
+from .utils import (REPO_ADMIN, REPO_MAINTAIN, REPO_READ, REPO_WRITE,
+                    RepoAccessLevel, RepoName, TeamSlug, Username,
+                    access_level_to_string)
+
+# Types for Terraform labels.
+# A "label" is that string that goes next to resource type or module name
+# in a Terraform block.
+# For example, in these blocks:
+#   resource "resourcetype" "blah1" { ... }
+#   module "blah2" { source = "./module/moduletype", ... }
+# the labels are "blah1" and "blah2".
+# Labels are used as the primary identifier for a Terraform block.
+# Per Terraform style, they should be lowercase and contain only [a-z0-9_],
+# so not dots, dashes, or spaces.
+RoleLabel = NewType("RoleLabel", str)
+RepoLabel = NewType("RepoLabel", str)
+UserLabel = NewType("UserLabel", str)
+
+
+@dataclass
+class TerraRole:
+    """
+    A role grants a set of repository permissions to a set of users.
+
+    Each of these roles will be represented as a Terraform module
+    invocation.
+    """
+
+    label: RoleLabel
+    repo_access: Dict[RepoName, RepoAccessLevel]
+    members: Set[Username]
+
+
+def main():
+    """
+    Script entrypoint
+    """
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    assert 2 <= len(sys.argv) <= 3, "Expected 1-2 args: ORGANIZATION [--phony]"
+
+    org_slug = sys.argv[1]
+    input_json_fpath = f"migrate/export-{org_slug}.json"
+    output_repos_fpath = f"repos_{org_slug}.tf"
+    output_users_fpath = f"users_{org_slug}.tf"
+    output_roles_fpath = f"roles_{org_slug}.tf"
+
+    if len(sys.argv) == 3:
+        assert sys.argv[2] == "--phony"
+        phony_repos = True
+    else:
+        phony_repos = False
+
+    # Load JSON generated by export_edx_to_json.py
+    export = get_export(input_json_fpath)
+    export_timestamp: str = export["export_date_utc"]
+    export_comment = (
+        f"// This file was automatically generated based on the contents of\n"
+        f"// the {org_slug!r} GitHub organization as of {export_timestamp}.\n"
+        "// It is likely to be reorganized, refactored, split, or combined with\n"
+        "// other Terraform files in the near future."
+    )
+
+    # Get map from teams to member lists.
+    team_members = {team["slug"]: team["members"] for team in export["teams"]}
+
+    # Calculate repo access granted to each team and user.
+    team_repo_access: Dict[TeamSlug, Dict[RepoName, RepoAccessLevel]] = defaultdict(
+        dict
+    )
+    user_repo_access: [Username, Dict[RepoName, RepoAccessLevel]] = defaultdict(dict)
+    for repo in export["repos"]:
+        repo_name = repo["name"]
+        # Recall that: all repos in the export are public.
+        # Since there's no reason we need to spell out read permissions for anyone,
+        # we drop any access grants that aren't at least "write" level.
```

Comment:
[curiosity]: Was this the optimization that sped everything up?

Reply:
It was to reduce the amount of noise in the resulting Terraform.